src/paths/floyd_warshall.c

```diff
@@ -140,3 +142,161 @@ igraph_error_t igraph_distances_floyd_warshall(
+        IGRAPH_ERROR("Invalid mode.", IGRAPH_EINVAL);
+    }
+
+    if (weights && igraph_vector_is_any_nan(weights)) {
+        IGRAPH_ERROR("Weight vector must not contain NaN values.", IGRAPH_EINVAL);
+    }
+
+    IGRAPH_CHECK(igraph_matrix_resize(res, no_of_nodes, no_of_nodes));
+    igraph_matrix_fill(res, IGRAPH_INFINITY);
+
+    for (igraph_integer_t v=0; v < no_of_nodes; v++) {
+        MATRIX(*res, v, v) = 0;
+    }
+
+    for (igraph_integer_t e=0; e < no_of_edges; e++) {
+        igraph_integer_t from = IGRAPH_FROM(graph, e);
+        igraph_integer_t to = IGRAPH_TO(graph, e);
+        igraph_real_t w = weights ? VECTOR(*weights)[e] : 1;
+
+        if (w < 0) {
+            if (mode == IGRAPH_ALL) {
+                IGRAPH_ERRORF("Negative edge weight (%g) found in undirected graph "
+                              "while calculating distances with Floyd-Warshall.",
+                              IGRAPH_ENEGLOOP, w);
+            } else if (to == from) {
+                IGRAPH_ERRORF("Self-loop with negative weight (%g) found "
+                              "while calculating distances with Floyd-Warshall.",
+                              IGRAPH_ENEGLOOP, w);
+            }
+        }
+
+        if (out && MATRIX(*res, from, to) > w) MATRIX(*res, from, to) = w;
+        if (in && MATRIX(*res, to, from) > w) MATRIX(*res, to, from) = w;
+    }
+
+    /* predecessors[u][v] is the second but last vertex on the shortest path from u to v */
+    igraph_matrix_int_t predecessors;
+    /* children[u][i] is the i-th children of u in a tree of shortest distances
+       rooted at k in the main loop below */
+    igraph_matrix_int_t children;
+    /* no_of_children[u] is the number of children of u
+       rooted at k in the main loop below */
+    igraph_vector_int_t no_of_children;
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&no_of_children, no_of_nodes);
+
+    igraph_vector_int_t dfs_traversal;
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&dfs_traversal, no_of_nodes);
+
+    igraph_vector_int_t dfs_skip;
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&dfs_skip, no_of_nodes);
+
+    igraph_stack_int_t stack;
+    IGRAPH_STACK_INT_INIT_FINALLY(&stack, no_of_nodes);
+
+    IGRAPH_CHECK(igraph_matrix_int_init(&predecessors, no_of_nodes, no_of_nodes));
+    IGRAPH_FINALLY(igraph_matrix_int_destroy, &predecessors);
+
+    IGRAPH_CHECK(igraph_matrix_int_init(&children, no_of_nodes, no_of_nodes));
+    IGRAPH_FINALLY(igraph_matrix_int_destroy, &children);
+
+    for (igraph_integer_t v=0; v < no_of_nodes; v++) {
+        for (igraph_integer_t u=0; u < no_of_nodes; u++) {
+            MATRIX(predecessors, u, v) = u;
+        }
+    }
+
+    for (igraph_integer_t k=0; k < no_of_nodes; k++) {
+        /* resetting no_of_children vector */
+        for (igraph_integer_t parent=0; parent < no_of_nodes; parent++)
+            VECTOR(no_of_children)[parent] = 0;
+        /* constructing the tree out_k (as in the paper) but
+           representing it as the children matrix */
```

Comment:
Representing the tree by a matrix gives a substantial improvement in comparison with a linear (space) tree:

|  3 vcount=100, p=0.5, Unweighted Floyd-Warshall-tree-speedup-linear-tree, 200x   0.746s  0.746s      0s
|  3 vcount=100, p=0.5, Unweighted Floyd-Warshall-tree-speedup, 200x               0.468s  0.468s      0s
|  4 vcount=100, p=0.5, Floyd-Warshall-tree-speedup-linear-tree, 200x              0.647s  0.647s      0s
|  4 vcount=100, p=0.5, Floyd-Warshall-tree-speedup, 200x                          0.394s  0.394s      0s
|  7 vcount=100, p=0.5, Floyd-Warshall-tree-speedup-linear-tree (negative), 200x   0.657s  0.657s      0s
|  7 vcount=100, p=0.5, Floyd-Warshall-tree-speedup (negative), 200x               0.413s  0.413s      0s

|  3 vcount=500, p=0.1, Unweighted Floyd-Warshall-tree-speedup-linear-tree, 10x     1.24s   1.24s      0s
|  3 vcount=500, p=0.1, Unweighted Floyd-Warshall-tree-speedup, 10x                0.798s  0.798s      0s
|  4 vcount=500, p=0.1, Floyd-Warshall-tree-speedup-linear-tree, 10x                1.27s   1.27s      0s
|  4 vcount=500, p=0.1, Floyd-Warshall-tree-speedup, 10x                           0.933s  0.933s      0s
|  7 vcount=500, p=0.1, Floyd-Warshall-tree-speedup-linear-tree  (negative), 10x    1.26s   1.26s      0s
|  7 vcount=500, p=0.1, Floyd-Warshall-tree-speedup (negative), 10x                0.934s  0.924s   0.01s

|  3 vcount=1500, p=0.02, Unweighted Floyd-Warshall-tree-speedup-linear-tree, 1x    2.69s   2.68s   0.01s
|  3 vcount=1500, p=0.02, Unweighted Floyd-Warshall-tree-speedup, 1x                2.22s   2.21s   0.01s
|  4 vcount=1500, p=0.02, Floyd-Warshall-tree-speedup-linear-tree, 1x                3.4s    3.4s      0s
|  4 vcount=1500, p=0.02, Floyd-Warshall-tree-speedup, 1x                           2.96s   2.96s      0s
|  7 vcount=1500, p=0.02, Floyd-Warshall-tree-speedup-linear-tree (negative), 1x    3.34s   3.34s      0s
|  7 vcount=1500, p=0.02, Floyd-Warshall-tree-speedup (negative), 1x                3.02s   3.02s      0s

Reply:
The distance matrix itself already takes up $O(|V|^2)$ storage, so I think we don't need to worry about the memory use of this approach.